src/comm.cc

```diff
@@ -226,6 +244,10 @@ comm_open(int sock_type,
           int flags,
           const char *note)
 {
+    // assume zero-port callers do not need to know the assigned port right away
+    if ((flags & COMM_DOBIND) && sock_type == SOCK_STREAM && addr.port() == 0)
```

Comment:
The current PR condition does not work as expected because some popular callers do _not_ set COMM_DOBIND even though they do want Comm to bind() to the address. The following condition does work in my quick tests:

```suggestion
    if (sock_type == SOCK_STREAM && addr.port() == 0 && !addr.isAnyAddr())
```

... but please review whether further adjustments are needed to cover all the cases we want to cover (and only them).

Reply:
If we want to apply IP_BIND_ADDRESS_NO_PORT whenever Comm calls commBind() with zero TCP port, then this condition should probably become

```suggestion
    if (sock_type == SOCK_STREAM && addr.port() == 0 && ((flags & COMM_DOBIND) || !addr.isAnyAddr()))
```

Compared to the previous [suggestion](https://github.com/squid-cache/squid/pull/1115#discussion_r943464151), the above condition also covers use cases where the caller did not give a _specific_ IP address to bind to but did request Comm to bind (presumably to some address of OS choosing). I do not know whether such use cases exist now, but I suspect then may appear in the future even if they do not exist today.

The official condition for calling commBind() is

```C++
    if ( (flags & COMM_DOBIND) || addr.port() > 0 || !addr.isAnyAddr() ) {
```

Reply:
Just to avoid misunderstanding: I am not saying the code is buggy just because the code "looks" wrong. I did test the latest branch code (6cfbf23), and it does _not_ run IP_BIND_ADDRESS_NO_PORT-setting code (at all) when Squid is configured with a tcp_outgoing_address. That code does not run because Squid does _not_ set `COMM_DOBIND` in that case.

Reply:
FTR these are the cases (feel free to edit this table as needed):
<table>
<tr><th>case</th><th>address</th><th>port</th><th>DOBIND</th><th>REUSEADDR</th><th>IP_BIND_ADDRESS_NO_PORT</th></tr>

<tr><th>listen-port</th><td>NO</td><td>non-0</td><td>YES</td><td>NO</td><td>NO</td></tr>
<tr><th>listen-host:port</th><td>YES</th><td>non-0</td><td>YES</td><td>NO</td><td>NO</td></tr>

<tr><th>tcp_outgoing_address</th><td>YES</td><td>0</td><td>YES</td><td>NO</td><td>YES</td></tr>

<tr><th>FTP gateway DATA</th><td>YES</td><td>0</td><td>NO</td><td>NO</td><td>YES</td></tr>
<tr><th>FTP gateway PORT fallback</th><td>YES</td><td>non-0</td><td>NO</td><td>NO</td><td>NO</td></tr>
<tr><th>FTP gateway PASV</th><td>YES</td><td>non-0</td><td>NO</td><td>NO</td><td>NO</td></tr>
<tr><th>FTP relay DATA</th><td>YES</td><td>0</td><td>YES</td><td>NO</td><td>YES</td></tr>

<tr><th>DNS TCP</th><td>YES</td><td>0</td><td>YES</td><td>NO</td><td>YES</td></tr>

</table>

Reply:
I committed the second proposed suggestion (branch commit [d96d055](https://github.com/squid-cache/squid/pull/1115/commits/d96d055e52a5b91900540941a2d172c830105b23)). This commit addresses my change request, so I am resolving this thread.

> FTR these are the cases (feel free to edit this table as needed):

FWIW, I have not audited the proposed table, but it does not seem to contradict the committed fix AFAICT. Going forward, I suspect that commBind() triggers should be revised, but that work is outside _this_ PR scope.